modules/renter/hostdb/scan_test.go

```diff
@@ -0,0 +1,156 @@
+	if _, ok := hdb.activeHosts[h.NetAddress]; ok {
+		t.Error("decrementing did not remove host from activeHosts")
+	}
+
+	// Decrement reliability to 0. This should remove the host from allHosts.
+	hdb.decrementReliability(h.NetAddress, h.reliability)
+	if _, ok := hdb.allHosts[h.NetAddress]; ok {
+		t.Error("decrementing did not remove host from allHosts")
+	}
+}
+
+// probeDialer is used to test the threadedProbeHosts method. A simple type
+// alias is used so that it can easily be redefined during testing, allowing
+// multiple behaviors to be tested.
+type probeDialer func(modules.NetAddress, time.Duration) (net.Conn, error)
+
+func (dial probeDialer) DialTimeout(addr modules.NetAddress, timeout time.Duration) (net.Conn, error) {
+	return dial(addr, timeout)
+}
+
+// TestThreadedProbeHosts tests the threadedProbeHosts method.
+func TestThreadedProbeHosts(t *testing.T) {
+	hdb := bareHostDB()
+
+	// create a host to send to threadedProbeHosts
+	h := new(hostEntry)
+	h.NetAddress = "foo"
+	h.reliability = baseWeight // enough to withstand a few failures
+
+	// define a helper function for running threadedProbeHosts. We send the
+	// hostEntry, close the channel, and then call threadedProbeHosts.
+	// threadedProbeHosts will receive the host, loop once, and return after
+	// seeing the channel has closed.
+	runProbe := func(h *hostEntry) {
+		hdb.scanPool <- h
+		close(hdb.scanPool)
+		hdb.threadedProbeHosts()
+		// reset hdb.scanPool
+		hdb.scanPool = make(chan *hostEntry, 1)
+	}
+
+	// make the dial fail
+	hdb.dialer = probeDialer(func(modules.NetAddress, time.Duration) (net.Conn, error) {
+		return nil, net.UnknownNetworkError("fail")
+	})
+	runProbe(h)
+	if _, ok := hdb.activeHosts[h.NetAddress]; ok {
+		t.Error("unresponsive host was added")
+	}
+
+	// make the RPC fail
+	hdb.dialer = probeDialer(func(modules.NetAddress, time.Duration) (net.Conn, error) {
+		ourPipe, theirPipe := net.Pipe()
+		ourPipe.Close()
+		return theirPipe, nil
+	})
+	runProbe(h)
+	if _, ok := hdb.activeHosts[h.NetAddress]; ok {
+		t.Error("unresponsive host was added")
+	}
+
+	// normal host
+	hdb.dialer = probeDialer(func(modules.NetAddress, time.Duration) (net.Conn, error) {
+		// create an in-memory conn and spawn a goroutine to handle our half
+		ourConn, theirConn := net.Pipe()
+		go func() {
+			// read the RPC
+			encoding.ReadObject(ourConn, new(types.Specifier), types.SpecifierLen)
+			// write old host settings
+			encoding.WriteObject(ourConn, oldHostSettings{
+				NetAddress: "probed",
+			})
+			ourConn.Close()
+		}()
+		return theirConn, nil
+	})
+	runProbe(h)
+	if _, ok := hdb.activeHosts[h.NetAddress]; !ok {
+		t.Error("host was not added")
+	}
+
+	// TODO: respond with old host settings
+}
+
+// TestThreadedScan tests the threadedScan method.
+func TestThreadedScan(t *testing.T) {
+	hdb := bareHostDB()
+	// use a real sleeper; this will prevent threadedScan from looping too
+	// quickly.
+	hdb.sleeper = stdSleeper{}
```

Comment:
you could also send in a sleeper that sleeps for like 1/3 time or something. Not sure that it's needed here, but it's a thought.